ipn/localapi/localapi.go

```diff
@@ -637,21 +637,32 @@ func (h *Handler) serveDebugPortmap(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	var logLock sync.Mutex
+	var (
+		logLock     sync.Mutex
+		handlerDone bool
+	)
 	logf := func(format string, args ...any) {
 		if !strings.HasSuffix(format, "\n") {
 			format = format + "\n"
 		}
 
 		logLock.Lock()
 		defer logLock.Unlock()
+		if handlerDone {
```

Comment:
I feel like something near here deserves a comment, it took me a while to figure out how we could be calling this from another goroutine (which I guess is from the callback to portmapper.NewClient, but that was not at all obvious).

Reply:
Something along the lines of: "The callback may try to log after the request returns, and at that point w is invalid." or words to that effect.

Reply:
Good call; added a comment with an explanation.